infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -0,0 +1,133 @@
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Sequencer server cache.
+ * Contains transaction conflict-resolution data structures.
+ */
+@Slf4j
+public class SequencerServerCache {
+    /**
+     * TX conflict-resolution information:
+     * <p>
+     * a cache of recent conflict keys and their latest global-log position.
+     */
+    private final Cache<String, Long> conflictToGlobalTailCache;
+
+    /**
+     * A "wildcard" representing the maximal update timestamp of
+     * all the conflict keys which were evicted from the cache
+     */
+    @Getter
+    private long maxConflictWildcard = Address.NOT_FOUND;
+
+    /**
+     * maxConflictNewSequencer represents the max update timestamp of all the conflict keys
+     * which were evicted from the cache by the time this server is elected
+     * the primary sequencer. This means that any snapshot timestamp below this
+     * actual threshold would abort due to NEW_SEQUENCER cause.
+     */
+    @Getter
+    private long maxConflictNewSequencer = Address.NOT_FOUND;
+
+    @VisibleForTesting
+    public SequencerServerCache(long cacheSize, CacheWriter<String, Long> writer) {
+        this.conflictToGlobalTailCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                .writer(writer)
+                .executor(Runnable::run)
+                .recordStats()
+                .build();
+    }
+
+    /**
+     * Limited cache by size.
+     * Eviction policy described here https://github.com/ben-manes/caffeine/wiki/Writer
+     * @param cacheSize cache size
+     */
+    public SequencerServerCache(long cacheSize) {
+        CacheWriter<String, Long> writer = new CacheWriter<String, Long>() {
+            @Override
+            public void write(@Nonnull String key, @Nonnull Long value) {
+                //ignore
+            }
+
+            @Override
+            public void delete(@Nonnull String key, @Nullable Long globalAddress, @Nonnull RemovalCause cause) {
+                if (cause == RemovalCause.REPLACED) {
+                    return;
+                }
+
+                log.trace(
+                        "Updating maxConflictWildcard. Old = '{}', new ='{}' conflictParam = '{}'. Cause = '{}'",
+                        maxConflictWildcard, globalAddress, key, cause
+                );
+
+                if (globalAddress == null) {
+                    globalAddress = Address.NOT_FOUND;
+                }
+                maxConflictWildcard = Math.max(globalAddress, maxConflictWildcard);
+            }
+        };
+
+        this.conflictToGlobalTailCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                .executor(Runnable::run)
+                .writer(writer)
+                .recordStats()
+                .build();
+    }
+
+    public Long getIfPresent(String conflictKeyHash) {
+        return conflictToGlobalTailCache.getIfPresent(conflictKeyHash);
+    }
+
+    public void invalidate(long trimMark) {
```

Comment:
maybe rename this to invalidateUpToTrimMark and please add a comment.

Reply:
Why repeat same thing twice? We invalidate the cache, also we see trim mark in a parameter. Making method names more verbose trying to explain what they do, it seems not so good idea to me. Also, we can explain details in javadoc

Reply:
a comment has added